SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/RPCService.cpp

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/RPCs/RPCService.h"
+
+namespace SpatialGDK
+{
+RPCService::RPCService(const FSubView& InRemoteSubView, const FSubView& InLocalAuthSubView)
+	: RemoteSubView(&InRemoteSubView)
+	, LocalAuthSubView(&InLocalAuthSubView)
+{
+}
+
+void RPCService::AddRPCQueue(FName QueueName, RPCQueueDescription&& Desc)
+{
+	Queues.Add(QueueName, MoveTemp(Desc));
+}
+
+void RPCService::AddRPCReceiver(FName ReceiverName, RPCReceiverDescription&& Desc)
+{
+	Receivers.Add(ReceiverName, MoveTemp(Desc));
+}
+
+void RPCService::AdvanceSenderQueues()
+{
+	for (const EntityDelta& Delta : LocalAuthSubView->GetViewDelta().EntityDeltas)
+	{
+		switch (Delta.Type)
+		{
+		case EntityDelta::UPDATE:
+		{
+			for (const ComponentChange& Change : Delta.ComponentUpdates)
+			{
+				RPCReadingContext Ctx;
+				Ctx.EntityId = Delta.EntityId;
+				Ctx.ComponentId = Change.ComponentId;
+				Ctx.Update = Change.Update;
+				Ctx.Fields = Schema_GetComponentUpdateFields(Change.Update);
+
+				for (auto& QueueEntry : Queues)
+				{
+					if (QueueEntry.Value.Sender->GetComponentsToReadOnUpdate().Contains(Ctx.ComponentId))
+					{
+						QueueEntry.Value.Sender->OnUpdate(Ctx);
+					}
+				}
+			}
+			for (const ComponentChange& Change : Delta.ComponentsRefreshed)
+			{
+				RPCReadingContext Ctx;
+				Ctx.EntityId = Delta.EntityId;
+				Ctx.ComponentId = Change.ComponentId;
+				Ctx.Fields = Schema_GetComponentDataFields(Change.CompleteUpdate.Data);
+
+				for (auto& QueueEntry : Queues)
+				{
+					if (QueueEntry.Value.Sender->GetComponentsToReadOnUpdate().Contains(Ctx.ComponentId))
+					{
+						QueueEntry.Value.Sender->OnUpdate(Ctx);
+					}
+				}
+			}
+			break;
+		}
+		case EntityDelta::ADD:
+		{
+			const EntityViewElement& ViewElement = LocalAuthSubView->GetView().FindChecked(Delta.EntityId);
+			for (auto& QueueEntry : Queues)
+			{
+				if (ViewElement.Authority.Contains(QueueEntry.Value.Authority))
+				{
+					QueueEntry.Value.Queue->OnAuthGained(Delta.EntityId, ViewElement);
+					QueueEntry.Value.Sender->OnAuthGained(Delta.EntityId, ViewElement);
+				}
+			}
+		}
+		break;
+		case EntityDelta::REMOVE:
+			for (auto& QueueEntry : Queues)
+			{
+				QueueEntry.Value.Queue->OnAuthLost(Delta.EntityId);
+				QueueEntry.Value.Sender->OnAuthLost(Delta.EntityId);
+			}
+			break;
+		case EntityDelta::TEMPORARILY_REMOVED:
+		{
+			const EntityViewElement& ViewElement = LocalAuthSubView->GetView().FindChecked(Delta.EntityId);
+			for (auto& QueueEntry : Queues)
+			{
+				QueueEntry.Value.Queue->OnAuthLost(Delta.EntityId);
+				QueueEntry.Value.Sender->OnAuthLost(Delta.EntityId);
+				if (ViewElement.Authority.Contains(QueueEntry.Value.Authority))
```

Comment:
Question: is this a normal / expected occurrence that in some case a `TEMPORARILY_REMOVED` only does the lost but not the gained part?

Reply:
For this specific view, I think not, because presence in the view is tied to authority.
For the other view, it does not assume auth for the presence, so the entity could disappear with some auth, are reappear with a different auth.